src/test/java/teammates/test/cases/storage/StudentsDbTest.java

```diff
@@ -285,8 +357,76 @@ public void testDeleteStudent() throws InvalidParametersException, EntityDoesNot
         
         studentsDb.deleteStudent(s.course, s.email);
 
-      //Untested case: The deletion is not persisted immediately (i.e. persistence delay)
-      //       Reason: Difficult to reproduce a persistence delay during testing
+        ______TS("Can delete old student entity without a CourseStudent copy");
+        
+        StudentAttributes oldStudent = createOldStudentAttributes("deleteStudent");
+        assertTrue("Old student entity should be created", isOldStudentExists(oldStudent));
+        assertFalse("New student entity should not be created", isNewStudentExists(oldStudent));
+        
+        studentsDb.deleteStudent(oldStudent.course, oldStudent.email);
+        assertNull(studentsDb.getStudentForEmail(oldStudent.course, oldStudent.email));
+        assertFalse(isOldStudentExists(oldStudent));
+        
+        ______TS("delete works for a student having both Student and CourseStudent");
+        oldStudent = createOldStudentAttributes("deleteStudent");
+        copyOldStudentEntityToCourseStudent(oldStudent.email, oldStudent.course);
+        
+        assertTrue(isOldStudentExists(oldStudent));
+        assertTrue(isNewStudentExists(oldStudent));
+        
+        studentsDb.deleteStudent(oldStudent.course, oldStudent.email);
+        assertNull(studentsDb.getStudentForEmail(oldStudent.course, oldStudent.email));
+        
+        assertFalse(isOldStudentExists(oldStudent));
+        assertFalse(isNewStudentExists(oldStudent));
+        
+    }
+    
+    // This method can be removed after all Students have been migrated to CourseStudents
+    private StudentAttributes createOldStudentAttributes(String testName)
+            throws InvalidParametersException {
+        StudentAttributes s = new OldStudentEntityPersistenceAttributes();
+        s.name = "valid student";
+        s.course = "valid-course" + testName;
+        s.email = "validOldStudent" + testName + "@email.com";
+        s.team = "validTeamName";
+        s.section = "validSectionName";
+        s.comments = "";
+        s.googleId = "";
+        
+        studentsDb.createEntityWithoutExistenceCheck(s);
+        
+        return studentsDb.getStudentForEmail(s.course, s.email);
+    }
+    
+    private StudentAttributes copyOldStudentEntityToCourseStudent(String email, String course)
+            throws InvalidParametersException {
+        studentsDb.copyStudentToCourseStudent(course, email);
+        return studentsDb.getStudentForEmail(course, email);
+    }
+    
+    @SuppressWarnings("deprecation")
+    // This method can be removed after all Students have been migrated to CourseStudents
+    private boolean isOldStudentExists(StudentAttributes s) {
+        boolean isOldStudentExist = false;
+        for (StudentAttributes studentsInDb : studentsDb.getAllOldStudents()) {
+            if (studentsInDb.getId().equals(s.getId())) {
+                isOldStudentExist = true;
+            }
+        }
+        return isOldStudentExist;
```

Comment:
`return false` here and do an early `return true` in the for-loop. Similarly for `isNewStudentExists`.